### PR TITLE
textparse: reject duplicate labels

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by brian-brazil** [comment]
How does the labels package deal with this?

**Comment by roidelapluie** [comment]
I have researched it.

Unless you use a Builder, labels is happy with duplicate labels (and should be, as it can be used in promql).

In this case, we do
```
var res labels
res.Metric(&res)
```
so we don't check for duplicate and ingest that promptly.

File: pkg/textparse/openmetricsparse_test.go [hunk 1]
```diff
@@ -442,6 +442,10 @@ func TestOpenMetricsParseErrors(t *testing.T) {
 			input: "empty_label_name{=\"\"} 0\n# EOF\n",
 			err:   "expected label name or left brace, got \"EQUAL\"",
 		},
+		{
```
Comment (roidelapluie) [inline pkg/textparse/openmetricsparse_test.go]: Note: we did not test that part of the code parser yet.

**Comment by roidelapluie** [comment]
```
benchmark                                                                    old ns/op     new ns/op     delta
BenchmarkParse/prometheus/no-decode-metric/promtestdata.txt-8                268           343           +27.99%
BenchmarkParse/prometheus/decode-metric/promtestdata.txt-8                   423           521           +23.17%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.txt-8             361           439           +21.61%
BenchmarkParse/expfmt-text/promtestdata.txt-8                                2765          3020          +9.22%
BenchmarkParse/prometheus/no-decode-metric/promtestdata.nometa.txt-8         203           310           +52.71%
BenchmarkParse/prometheus/decode-metric/promtestdata.nometa.txt-8            372           457           +22.85%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.nometa.txt-8      288           367           +27.43%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt-8                         1934          1953          +0.98%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.txt-8               341           419           +22.87%
BenchmarkParse/openmetrics/decode-metric/promtestdata.txt-8                  507           581           +14.60%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.txt-8            425           500           +17.65%
BenchmarkParse/expfmt-text/promtestdata.txt#01-8                             2888          3040          +5.26%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.nometa.txt-8        285           356           +24.91%
BenchmarkParse/openmetrics/decode-metric/promtestdata.nometa.txt-8           440           527           +19.77%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.nometa.txt-8     362           439           +21.27%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt#01-8                      1967          1983          +0.81%
BenchmarkGzip/promtestdata.txt-8                                             329           321           -2.43%
BenchmarkGzip/promtestdata.nometa.txt-8                                      214           211           -1.40%

benchmark                                                                    old MB/s          new MB/s          speedup
BenchmarkParse/prometheus/no-decode-metric/promtestdata.txt-8                1394948633.69     842444869.99      0.60x
BenchmarkParse/prometheus/decode-metric/promtestdata.txt-8                   526063350.10      353477776.50      0.67x
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.txt-8             814517192.63      516741131.51      0.63x
BenchmarkParse/expfmt-text/promtestdata.txt-8                                12766697.91       10627597.29       0.83x
BenchmarkParse/prometheus/no-decode-metric/promtestdata.nometa.txt-8         1840414568.37     783069439.58      0.43x
BenchmarkParse/prometheus/decode-metric/promtestdata.nometa.txt-8            504716207.02      368406565.32      0.73x
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.nometa.txt-8      885013202.31      563186441.53      0.64x
BenchmarkParse/expfmt-text/promtestdata.nometa.txt-8                         19556067.54       17421921.63       0.89x
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.txt-8               836418302.23      565854410.94      0.68x
BenchmarkParse/openmetrics/decode-metric/promtestdata.txt-8                  365256416.19      285607287.47      0.78x
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.txt-8            548421286.67      401016004.27      0.73x
BenchmarkParse/expfmt-text/promtestdata.txt#01-8                             10720815.28       9218074.78        0.86x
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.nometa.txt-8        875254680.59      563736416.81      0.64x
BenchmarkParse/openmetrics/decode-metric/promtestdata.nometa.txt-8           379476883.06      274503665.99      0.72x
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.nometa.txt-8     567810971.53      396471735.39      0.70x
BenchmarkParse/expfmt-text/promtestdata.nometa.txt#01-8                      20936698.20       18666887.68       0.89x
BenchmarkGzip/promtestdata.txt-8                                             862634374.30      918430722.57      1.06x
BenchmarkGzip/promtestdata.nometa.txt-8                                      1635152113.49     1716821679.98     1.05x

benchmark                                                                    old allocs     new allocs     delta
BenchmarkParse/prometheus/no-decode-metric/promtestdata.txt-8                0              0              +0.00%
BenchmarkParse/prometheus/decode-metric/promtestdata.txt-8                   2              2              +0.00%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.txt-8             1              1              +0.00%
BenchmarkParse/expfmt-text/promtestdata.txt-8                                30             30             +0.00%
BenchmarkParse/prometheus/no-decode-metric/promtestdata.nometa.txt-8         0              0              +0.00%
BenchmarkParse/prometheus/decode-metric/promtestdata.nometa.txt-8            2              2              +0.00%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.nometa.txt-8      1              1              +0.00%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt-8                         24             24             +0.00%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.txt-8               1              1              +0.00%
BenchmarkParse/openmetrics/decode-metric/promtestdata.txt-8                  3              3              +0.00%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.txt-8            2              2              +0.00%
BenchmarkParse/expfmt-text/promtestdata.txt#01-8                             30             30             +0.00%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.nometa.txt-8        1              1              +0.00%
BenchmarkParse/openmetrics/decode-metric/promtestdata.nometa.txt-8           3              3              +0.00%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.nometa.txt-8     2              2              +0.00%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt#01-8                      24             24             +0.00%
BenchmarkGzip/promtestdata.txt-8                                             0              0              +0.00%
BenchmarkGzip/promtestdata.nometa.txt-8                                      0              0              +0.00%

benchmark                                                                    old bytes     new bytes     delta
BenchmarkParse/prometheus/no-decode-metric/promtestdata.txt-8                0             28            +Inf%
BenchmarkParse/prometheus/decode-metric/promtestdata.txt-8                   144           171           +18.75%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.txt-8             48            75            +56.25%
BenchmarkParse/expfmt-text/promtestdata.txt-8                                1329          1329          +0.00%
BenchmarkParse/prometheus/no-decode-metric/promtestdata.nometa.txt-8         0             28            +Inf%
BenchmarkParse/prometheus/decode-metric/promtestdata.nometa.txt-8            144           171           +18.75%
BenchmarkParse/prometheus/decode-metric-reuse/promtestdata.nometa.txt-8      48            75            +56.25%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt-8                         922           922           +0.00%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.txt-8               39            67            +71.79%
BenchmarkParse/openmetrics/decode-metric/promtestdata.txt-8                  183           210           +14.75%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.txt-8            87            114           +31.03%
BenchmarkParse/expfmt-text/promtestdata.txt#01-8                             1329          1329          +0.00%
BenchmarkParse/openmetrics/no-decode-metric/promtestdata.nometa.txt-8        39            67            +71.79%
BenchmarkParse/openmetrics/decode-metric/promtestdata.nometa.txt-8           183           210           +14.75%
BenchmarkParse/openmetrics/decode-metric-reuse/promtestdata.nometa.txt-8     87            114           +31.03%
BenchmarkParse/expfmt-text/promtestdata.nometa.txt#01-8                      922           921           -0.11%
BenchmarkGzip/promtestdata.txt-8                                             417           417           +0.00%
BenchmarkGzip/promtestdata.nometa.txt-8                                      257           257           +0.00%
```

**Comment by brian-brazil** [comment]
> We probably want a change like this in tsdb too

Yes, that should be the core check like how we also drop empty labels.

**Comment by brian-brazil** [comment]
> labels is happy with duplicate labels (and should be, as it can be used in promql).

Why would promql need duplicate label names?

**Comment by roidelapluie** [comment]
> > labels is happy with duplicate labels (and should be, as it can be used in promql).
> 
> Why would promql need duplicate label names?

because {a="a",a="b"} is a valid promql query.

**Comment by brian-brazil** [comment]
> because {a="a",a="b"} is a valid promql query. And {a="a",a="a"} is even a query that can produce a result.

Those are selectors, not labels :)

File: pkg/textparse/promparse.go [hunk 2]
```diff
@@ -370,12 +371,20 @@ func (p *PromParser) Next() (Entry, error) {
...
 func (p *PromParser) parseLVals() error {
+	labels := [][]byte{}
 	t := p.nextToken()
 	for {
 		switch t {
 		case tBraceClose:
 			return nil
 		case tLName:
+			label := p.l.b[p.l.start:p.l.i]
+			for _, l := range labels {
```
Comment (brian-brazil) [inline pkg/textparse/promparse.go]: This is n^2, probably better to do it in the scrape code.

Comment (roidelapluie) [reply pkg/textparse/promparse.go]: if we do in the scrape code, our parser will not be open metrics complicant and will not be reusable by others. WDYT?

Comment (roidelapluie) [reply pkg/textparse/promparse.go]: I could extract that in a common function between the two parsers.

Comment (roidelapluie) [reply pkg/textparse/promparse.go]: It is also hard/not that mych added value to share that code between parsers as it is using private apis.

Comment (brian-brazil) [reply pkg/textparse/promparse.go]: This is our internal parser for speed, not an official OM parser. It's not meant to be reused by others, as it's tightly coupled to the scrape code.

**Comment by brian-brazil** [comment]
We'll also need to check what we get back from remote read. And figure out what to do about existing data like this.

File: pkg/textparse/promparse.go (hunk 2 above)
Comment (roidelapluie) [reply pkg/textparse/promparse.go]: I understand, and I did not know it. I was hoping that our om parser could be reused by 3rd party implementations. Also getting the code here will help when we will implement the openmetrics reference test suite.

Comment (brian-brazil) [reply pkg/textparse/promparse.go]: The reference test suite is in Python.

This is never intended to be a fully correct OM parser that detects all errors, just a good enough parser that will hopefully catch the main errors.

**Comment by mback2k** [comment]
> And figure out what to do about existing data like this.

This is important, because at least I now have existing data with duplicate labels which I cannot get rid of easily. :-(

**Comment by brian-brazil** [comment]
We can't fix such data, we have to choose exactly how much of it we throw away automatically - up to whole blocks.

**Comment by mback2k** [comment]
What does "whole blocks" mean? A timeframe? A timeseries? The specific metric affected?

I would like to propose the following order of approaches to duplicate labels in old data:

1. Try to fix it by merging the labels if the values are equal.
2. Skip (error out?) the broken metric timeseries, if label values are unequal.

**Comment by brian-brazil** [comment]
> Skip (error out?) the broken metric timeseries, if label values are unequal.

Partial data isn't something you can really reason about, we should be dropping the whole block (time range).

**Comment by mback2k** [comment]
Sorry, I still not get it. Are you talking about dropping just the affected metric with the duplicate label or all metrics for the time range?

**Comment by brian-brazil** [comment]
All metrics in the time range. It's basically a corrupted block.

**Comment by roidelapluie** [comment]
#6664 is better than this.

**Comment by brian-brazil** [comment]
We still need a safety check in the remote read code I think

**Pull request closed without merging**